File: sdk/go/client_test.go
package synapse

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"strings"
	"testing"
)

type roundTrip func(*http.Request) (*http.Response, error)

func (f roundTrip) RoundTrip(r *http.Request) (*http.Response, error) { return f(r) }

func testClient(t *testing.T, handle roundTrip) *Client {
	t.Helper()
	c, err := New("http://test", Options{
		APIKey:     "sk_test",
		OrgID:      "11111111-1111-1111-1111-111111111111",
		HTTPClient: &http.Client{Transport: handle},
	})
	if err != nil {
		t.Fatal(err)
	}
	return c
}

func jsonResponse(status int, body string) *http.Response {
	return &http.Response{
		StatusCode: status,
		Header:     http.Header{"Content-Type": []string{"application/json"}},
		Body:       io.NopCloser(strings.NewReader(body)),
	}
}

func TestAuthAndOrgHeaders(t *testing.T) {
	var gotAuth, gotOrg string
	c := testClient(t, func(r *http.Request) (*http.Response, error) {
		gotAuth = r.Header.Get("Authorization")
		gotOrg = r.Header.Get("X-Org-Id")
		return jsonResponse(200, `{"id":"u1"}`), nil
	})
	if _, err := c.Auth().Me(context.Background()); err != nil {
		t.Fatal(err)
	}
	if gotAuth != "Bearer sk_test" {
		t.Errorf("auth header = %q", gotAuth)
	}
	if gotOrg != "11111111-1111-1111-1111-111111111111" {
		t.Errorf("org header = %q", gotOrg)
	}
}

func TestConsumePayload(t *testing.T) {
	var body string
	c := testClient(t, func(r *http.Request) (*http.Response, error) {
		raw, _ := io.ReadAll(r.Body)
		body = string(raw)
		return jsonResponse(200, `{"total":5}`), nil
	})
	out, err := c.Usage().Consume(context.Background(), "api_requests", 5)
	if err != nil {
		t.Fatal(err)
	}
	if out["total"].(float64) != 5 {
		t.Errorf("total = %v", out["total"])
	}
	var parsed struct {
		Events []map[string]any `json:"events"`
	}
	if err := json.Unmarshal([]byte(body), &parsed); err != nil {
		t.Fatal(err)
	}
	if parsed.Events[0]["metric"] != "api_requests" || parsed.Events[0]["quantity"].(float64) != 5 {
		t.Errorf("payload = %s", body)
	}
}

func Test204ReturnsNoError(t *testing.T) {
	c := testClient(t, func(*http.Request) (*http.Response, error) {
		return &http.Response{StatusCode: 204, Body: io.NopCloser(strings.NewReader(""))}, nil
	})
	if err := c.APIKeys().Revoke(context.Background(), "k1"); err != nil {
		t.Errorf("revoke: %v", err)
	}
}

func TestLimitErrorTyped(t *testing.T) {
	c := testClient(t, func(*http.Request) (*http.Response, error) {
		return jsonResponse(402, `{"title":"usage limit exceeded","metric":"api_requests","limit":100}`), nil
	})
	_, err := c.Usage().Consume(context.Background(), "api_requests", 500)
	var limitErr *LimitError
	if !errors.As(err, &limitErr) {
		t.Fatalf("want LimitError, got %T: %v", err, err)
	}
	if limitErr.Metric() != "api_requests" {
		t.Errorf("metric = %q", limitErr.Metric())
	}
}

func TestFeatureGateErrorTyped(t *testing.T) {
	c := testClient(t, func(*http.Request) (*http.Response, error) {
		return jsonResponse(403, `{"title":"feature not entitled","feature":"advanced_reports","available_in":["pro"]}`), nil
	})
	_, err := c.Subscription().Change(context.Background(), "pro")
	var gateErr *FeatureGatedError
	if !errors.As(err, &gateErr) {
		t.Fatalf("want FeatureGatedError, got %T: %v", err, err)
	}
	if gateErr.Feature() != "advanced_reports" {
		t.Errorf("feature = %q", gateErr.Feature())
	}
	if len(gateErr.AvailableIn()) != 1 || gateErr.AvailableIn()[0] != "pro" {
		t.Errorf("available_in = %v", gateErr.AvailableIn())
	}
}

func TestRequiresCredentials(t *testing.T) {
	if _, err := New("http://test", Options{}); err == nil {
		t.Error("expected error for missing credentials")
	}
}

File: sdk/typescript/test/client.test.ts
import { describe, expect, it } from "vitest";
import {
  SynapseClient,
  SynapseAuthError,
  SynapseFeatureGatedError,
  SynapseLimitError,
  SynapseNotFoundError,
} from "../src/index.js";

type Recorded = { method: string; url: string; body?: string; auth: string | null; org: string | null };

function fakeFetch(responses: Array<{ status: number; body: unknown }>) {
  const requests: Recorded[] = [];
  let call = 0;
  const impl: typeof fetch = async (input, init) => {
    const url = String(input);
    requests.push({
      method: init?.method ?? "GET",
      url,
      body: typeof init?.body === "string" ? init.body : undefined,
      auth: (init?.headers as Record<string, string>)?.Authorization ?? null,
      org: (init?.headers as Record<string, string>)?.["X-Org-Id"] ?? null,
    });
    const preset = responses[Math.min(call, responses.length - 1)];
    call += 1;
    // 204 must carry an empty body — Response(null) makes .json() throw
    if (preset.status === 204) return new Response(null, { status: 204 });
    return new Response(JSON.stringify(preset.body), { status: preset.status });
  };
  return { impl, requests };
}

function client(fetchImpl: typeof fetch, orgId?: string) {
  return new SynapseClient("http://test", { apiKey: "sk_test", orgId, fetchImpl });
}

describe("requests", () => {
  it("sends auth + org headers", async () => {
    const { impl, requests } = fakeFetch([{ status: 200, body: { id: "u1" } }]);
    const c = client(impl, "11111111-1111-1111-1111-111111111111");
    await c.auth.me();
    expect(requests[0].auth).toBe("Bearer sk_test");
    expect(requests[0].org).toBe("11111111-1111-1111-1111-111111111111");
  });

  it("shapes consume payloads", async () => {
    const { impl, requests } = fakeFetch([{ status: 200, body: { total: 5 } }]);
    const c = client(impl);
    const result = (await c.usage.consume("api_requests", 5)) as { total: number };
    expect(result.total).toBe(5);
    expect(JSON.parse(requests[0].body!)).toEqual({
      events: [{ metric: "api_requests", quantity: 5 }],
    });
  });

  it("query params land in the URL", async () => {
    const { impl, requests } = fakeFetch([{ status: 200, body: {} }]);
    const c = client(impl);
    await c.usage.check("api_requests", 2);
    expect(requests[0].url).toContain("metric=api_requests");
    expect(requests[0].url).toContain("quantity=2");
  });

  it("204 returns undefined", async () => {
    const { impl } = fakeFetch([{ status: 204, body: null }]);
    const c = client(impl);
    expect(await c.apiKeys.revoke("k1")).toBeUndefined();
  });
});

describe("errors", () => {
  it("402 maps to SynapseLimitError with metric + limit", async () => {
    const { impl } = fakeFetch([
      { status: 402, body: { title: "usage limit exceeded", metric: "api_requests", limit: 100 } },
    ]);
    const c = client(impl);
    await expect(c.usage.consume("api_requests", 500)).rejects.toMatchObject({
      name: "SynapseLimitError",
      metric: "api_requests",
      limit: 100,
    });
  });

  it("403 feature gates carry available_in", async () => {
    const { impl } = fakeFetch([
      {
        status: 403,
        body: { title: "feature not entitled", feature: "advanced_reports", available_in: ["pro"] },
      },
    ]);
    const c = client(impl);
    try {
      await c.subscription.change("pro");
      expect.unreachable();
    } catch (err) {
      expect(err).toBeInstanceOf(SynapseFeatureGatedError);
      expect((err as SynapseFeatureGatedError).availableIn).toEqual(["pro"]);
    }
  });

  it("404 and 401 map to their types", async () => {
    const nf = fakeFetch([{ status: 404, body: { title: "not found" } }]);
    await expect(client(nf.impl).orgs.current()).rejects.toBeInstanceOf(SynapseNotFoundError);

    const au = fakeFetch([{ status: 401, body: { title: "unauthorized" } }]);
    await expect(client(au.impl).auth.me()).rejects.toBeInstanceOf(SynapseAuthError);
  });
});

describe("constructor", () => {
  it("requires credentials", () => {
    expect(() => new SynapseClient("http://test", {})).toThrow(/apiKey or accessToken/);
  });

  it("access token mode", async () => {
    const { impl, requests } = fakeFetch([{ status: 200, body: {} }]);
    const c = new SynapseClient("http://test", { accessToken: "jwt", fetchImpl: impl });
    await c.auth.me();
    expect(requests[0].auth).toBe("Bearer jwt");
  });
});
